crates/data-transformer/src/reverse_transformer/mod.rs

```diff
@@ -0,0 +1,69 @@
+mod transform;
+mod types;
+
+use crate::reverse_transformer::transform::{ReverseTransformer, TransformationError};
+use crate::shared::extraction::extract_function_from_selector;
+use starknet::core::types::contract::{AbiEntry, AbiFunction};
+use starknet_types_core::felt::Felt;
+
+#[derive(Debug, thiserror::Error)]
+pub enum ReverseTransformError {
+    #[error(r#"Function with selector "{0}" not found in ABI of the contract"#)]
+    FunctionNotFound(Felt),
+    #[error(transparent)]
+    TransformationError(#[from] TransformationError),
+}
+
+/// Transforms a calldata into a Cairo-like string representation of the arguments
+pub fn reverse_transform_input(
+    input: &[Felt],
+    abi: &[AbiEntry],
+    function_selector: &Felt,
+) -> Result<String, ReverseTransformError> {
+    let type_extractor = |function: AbiFunction| {
+        function
+            .inputs
+            .into_iter()
+            .map(|input| input.r#type)
+            .collect::<Vec<_>>()
+    };
+    reverse_transform(input, abi, function_selector, type_extractor)
+}
+
+/// Transforms a call output into a Cairo-like string representation of the return values
+pub fn reverse_transform_output(
+    output: &[Felt],
+    abi: &[AbiEntry],
+    function_selector: &Felt,
+) -> Result<String, ReverseTransformError> {
+    let type_extractor = |function: AbiFunction| {
+        function
+            .outputs
+            .into_iter()
+            .map(|input| input.r#type)
+            .collect::<Vec<_>>()
+    };
+    reverse_transform(output, abi, function_selector, type_extractor)
+}
+
+fn reverse_transform(
+    felts: &[Felt],
+    abi: &[AbiEntry],
+    function_selector: &Felt,
+    type_extractor: fn(AbiFunction) -> Vec<String>,
```

Comment:
I get the generalization of this with type_extractor, but maybe a less cleaver approach would be clearer to follow: Have this function accept argument types to convert directly and handle the argument extraction on the higher level?